prometheus_juju_backup_all_exporter/collector.py

```diff
@@ -1,169 +1,116 @@
-
-        def _fetch(self):
-            data = {
-                "some_metric": [
-                    {"labels": ["mongodb"], "value": 20},
-                    {"labels": ["juju-backup-all"], "value": 10}
-                ]
-            }
-            return data
-
-    exporter = Exporter(config.port)
-    exporter.register(
-        CustomCollector(
-            CustomGauge(
-                "custom_gauge_a",
-                "Example description of A",
-                labels=["app"],
-            ),
-        )
-    )
-    exporter.run()
-    ```
-    """
-
-    def __init__(self, *metrics: Any) -> None:
-        """Initialze the collector.
-
-        Args:
-            metrics: one or more custom metrics
-        """
-        self._metrics = metrics
-
-    @abstractmethod
-    def _fetch(self) -> Optional[Dict[str, List[Dict[str, Any]]]]:
-        """User defined _fetch method used by `self._update_metrics()`.
-
-        Return:
-            fetched data as metric samples
-
-        Examples:
-            returned_fetched_data = {
-                "some_metric": [
-                    {"label": [], "value": 10.0},
-                    {"label": ["foo"], "value": 20.0},
-                ]
-            }
-        """
-
-    def _update_metrics(self) -> None:
-        """Create new metrics based on existing metrics.
-
-        This is an internally used function to fetch new samples from the user
-        defined `_fetch()` method, and create new metrics based on the new
-        samples and the existing metric classes. The internal properties of
-        `_metrics` will be updated every time this internal function is called.
-        """
-        data = self._fetch()
-        new_metrics = []
-        for metric in self._metrics:
-            new_metric = metric.__class__(
-                metric.name,
-                metric.documentation,
-                labels=metric._labelnames,  # pylint: disable=W0212
-            )
-            new_metric.add_samples(data, metric.previous_data)
-            new_metrics.append(new_metric)
-        self._metrics = tuple(new_metrics)
-
-    def collect(self) -> Iterator[Metric]:
-        """Fetch data and update the internal metrics.
-
-        This is a callback method that is used internally within
-        `prometheus_client` every time the exporter server is queried. There is
-        not return values for this method but it needs to yield all the metrics.
-
-        Yields:
-            metrics: the internal metrics.
-        """
-        self._update_metrics()
-        for metric in self._metrics:
-            yield metric
-
-
-class BackupStateCollector(CollectorBase):  # pylint: disable=R0903
+class BackupStateCollector(SyncCollector):
     """Collector for backup state."""
 
-    def __init__(self, config: Config, *metrics: Metric) -> None:
-        """Initialize the class."""
-        super().__init__(*metrics)
-        self.config = config
+    @property
```

Comment:
is there a benefit of defining  `specifications` as property instead of regular attribute?

Reply:
The idea is to remind people to define what metrics they want, and not to forget it. Therefore, I also make it as abstract property, when users did not implement the `specifications` property, it will raise error.